### PR TITLE
perf(capabilities): Log capabilities providers that are slow

### PR DESCRIPTION
**Pull request by nickvergessen**
## Summary

Added a `sleep(3)` to test the log:
```
Capabilities of OCA\Talk\Capabilities took 3 seconds to generate.
```

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/CapabilitiesManager.php [hunk 1]
```diff
@@ -51,7 +51,16 @@ public function getCapabilities(bool $public = false, bool $initialState = false
 						// that we would otherwise inject to every page load
 						continue;
 					}
+					$startTime = microtime(true);
 					$capabilities = array_replace_recursive($capabilities, $c->getCapabilities());
+					$endTime = microtime(true);
+					$duration = $endTime - $startTime;
+					if ($duration > 0.3) {
+						$this->logger->warning('Capabilities of {className} took {duration} seconds to generate.', [
```
Comment (szaimen) [inline lib/private/CapabilitiesManager.php:59]:
```suggestion
						$this->logger->info('Capabilities of {className} took {duration} seconds to generate.', [
```

**Review by Altahrim — APPROVED** [review]

File: lib/private/CapabilitiesManager.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/CapabilitiesManager.php:59]: Then no one will see it from a real instance. But maybe we simply start with a warning at 1s or something and info below it

Comment (ChristophWurst) [reply lib/private/CapabilitiesManager.php:59]: You could also make it adaptive like https://github.com/nextcloud/server/blob/08cf60923c1ed8b41f4f0b9e34bea9a0ad00fd31/cron.php#L171-L177

**Review by ChristophWurst — APPROVED** [review]

**Review by szaimen — CHANGES_REQUESTED** [review]
See my comment

File: lib/private/CapabilitiesManager.php [hunk 2]
```diff
@@ -51,7 +57,27 @@ public function getCapabilities(bool $public = false, bool $initialState = false
 						// that we would otherwise inject to every page load
 						continue;
 					}
+					$startTime = microtime(true);
 					$capabilities = array_replace_recursive($capabilities, $c->getCapabilities());
+					$endTime = microtime(true);
+					$timeSpent = $endTime - $startTime;
+					if ($timeSpent > self::ACCEPTABLE_LOADING_TIME) {
```
Comment (ChristophWurst) [inline lib/private/CapabilitiesManager.php:64]: microtime gives you µs so we'll log anything that is slower than 100ns, right?

Comment (nickvergessen) [reply lib/private/CapabilitiesManager.php:64]: > microtime gives you µs

its float with seconds? https://3v4l.org/YF2v8

Comment (ChristophWurst) [reply lib/private/CapabilitiesManager.php:64]: Whooops. I read *returns the current Unix timestamp with microseconds* as *in microseconds*

**Review by ChristophWurst — APPROVED** [review]

**Review by szaimen — APPROVED** [review]
Fine by me but didnt test

**Pull request merged**